### PR TITLE
 testing: GrpcCleanupRule

### PR DESCRIPTION
**Pull request by dapengzhang0**
Added `GrpcCleanupRule` based on offline discussions with @ejona86 

This will ease a lot of test scenarios that we want to automatically shut down servers and channels, with much more flexibility such as that mentioned by #3624 than `GrpcServerRule`.

**ManagedChannel/Server cleanup details:**
- If the test has already failed, call `shutdownNow()` for each of the resources registered.
- If no error happens up to now, call `shutdown()` for each of the resources registered. If any error occurs, immediately call `shutdownNow()` for each resource.
- If no error happens up to now, call `awaitTermination()` with `timeout = deadline - current time` and assert termination for each resource.  If any error occurs, immediately call `shutdownNow()` for the current resource and all the rest resources.
- once the rule evaluation is done, throw the first exception encountered if any .

### CODE REVIEW

File: testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+   * Sets a positive total time limit for the automatic resource cleanup. If any of the resources
+   * registered to the rule fails to be released in time, the test will fail.
+   *
+   * <p>Note that the resource cleanup duration may or may not be counted as part of the JUnit
+   * {@link org.junit.rules.Timeout Timeout} rule's test duration, depending on which rule is
+   * applied first.
+   *
+   * @return this
+   */
+  public GrpcCleanupRule withTimeout(long timeout, TimeUnit timeUnit) {
```
Comment (ejona86) [inline testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: We tend to use `with` when a new instance will be returned. Let's use `set` or just `timeout` instead.

Looking at TemporaryFolder, it uses a builder instead. I'd be okay with that as well, if you'd prefer.

File: testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    timeoutNanos = timeUnit.toNanos(timeout);
+    return this;
+  }
+
+  /**
+   * Sets a specified time source for monitoring cleanup timeout.
+   *
+   * @return this
+   */
+  @SuppressWarnings("BetaApi") // Stopwatch.createUnstarted(Ticker ticker) is not Beta. Test only.
```
Comment (ejona86) [inline testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: What does "Stopwatch.createUnstarted(Ticker ticker) is not Beta." mean? It appears beta, as that's what you're suppressing...

File: testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+  public Statement apply(final Statement base, Description description) {
+    return new Statement() {
+      @Override
+      public void evaluate() throws Throwable {
+        try {
+          base.evaluate();
+        } catch (Throwable t) {
+          firstException = t;
+        } finally {
+          teardown();
```
Comment (ejona86) [inline testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: Why is this in a finally?

File: testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+    @Override
+    public void cleanUp() {
+      channel.shutdown();
+    }
+
+    @Override
+    public void forceCleanUp() {
+      try {
+        channel.shutdownNow();
+      } catch (Throwable ignorable) {
```
Comment (ejona86) [inline testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: shutdownNow should not throw. Let's not protect ourselves against that. If it happens, then we will just fix the bug.

File: testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    stopwatch.start();
+
+    for (int i = resources.size() - 1; i >= 0; i--) {
+      if (firstException != null) {
+        break;
+      }
+
+      try {
+        resources.get(i).cleanUp();
+      } catch (Throwable t) {
```
Comment (ejona86) [inline testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: shutdown should not throw. Let's not protect ourselves against that. If it happens, then we will just fix the bug.

File: testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+        if (!released) {
+          firstException = new AssertionError(
+              "Resource " + resources.get(i) + " can not be released in time at the end of test");
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        if (firstException == null) {
+          firstException = e;
+        }
+      } catch (Throwable t) {
```
Comment (ejona86) [inline testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: awaitTerminated should not throw arbitrary exceptions. Let's not protect ourselves against that. If it happens, then we will just fix the bug.

File: testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+        if (firstException == null) {
+          firstException = t;
+        }
+      } finally {
+        if (firstException != null) {
```
Comment (ejona86) [inline testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: This also seems unnecessary to be within a finally.

File: testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java (hunk 4 above)
Comment (dapengzhang0) [reply testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: The javadoc didn't say that. Even `ManagedChannelImpl` behaves so and javadoc says so, the user's test could be registering any `ManagedChannel` to the rule, say a mock `ManagedChannel` delegating to a `ManagedChannelImpl`, and their mock could happen to be really wild and evil. Same for `awaitTermination()` and `shutdown()`.

File: testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java (hunk 3 above)
Comment (dapengzhang0) [reply testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: Moved out of finally.

File: testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java (hunk 7 above)
Comment (dapengzhang0) [reply testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: Done.

File: testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java (hunk 1 above)
Comment (dapengzhang0) [reply testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: Done.

File: testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java (hunk 4 above)
Comment (ejona86) [reply testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: The methods do not define any exceptions will be thrown. Any Exception (vs Error) would be a bug (and as a general rule the amount of code that should avoid bugs in other code should be very limited) and any Error means "all hope is lost." If the user's test is doing something like that, I have no problem letting it fail. Those methods need to be reliable in general, and exposing such bugs seems fine to me. Garbage in, garbage out.

Comment (dapengzhang0) [reply testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: Done.

File: testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java (hunk 2 above)
Comment (dapengzhang0) [reply testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: I mean `Stopwatch` is not Beta, and it's API `Stopwatch.createUnstarted(Ticker ticker)` is also not marked with `@Beta`, even `Ticker` is marked with `@Beta` (and so I have to suppress).

Comment (ejona86) [reply testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: I still don't quite understand. Maybe "Ticker in Beta"?

Comment (dapengzhang0) [reply testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java]: Yes, Ticker is in Beta. I have to suprress.

**Pull request merged**